internal/controller/supervisorconfig/oidcclientwatcher/oidc_client_watcher.go

```diff
@@ -0,0 +1,317 @@
+func validateAllowedGrantTypes(oidcClient *v1alpha1.OIDCClient, conditions []*v1alpha1.Condition) []*v1alpha1.Condition {
+	switch {
+	case !allowedGrantTypesContains(oidcClient, authorizationCodeGrantTypeName):
+		conditions = append(conditions, &v1alpha1.Condition{
+			Type:   allowedGrantTypesValid,
+			Status: v1alpha1.ConditionFalse,
+			Reason: reasonMissingRequiredValue,
+			Message: fmt.Sprintf("%q must always be included in %q",
+				authorizationCodeGrantTypeName, allowedGrantTypesFieldName),
+		})
+	case allowedScopesContains(oidcClient, offlineAccessScopeName) && !allowedGrantTypesContains(oidcClient, refreshTokenGrantTypeName):
+		conditions = append(conditions, &v1alpha1.Condition{
+			Type:   allowedGrantTypesValid,
+			Status: v1alpha1.ConditionFalse,
+			Reason: reasonMissingRequiredValue,
+			Message: fmt.Sprintf("%q must be included in %q when %q is included in %q",
+				refreshTokenGrantTypeName, allowedGrantTypesFieldName, offlineAccessScopeName, allowedScopesFieldName),
+		})
+	case allowedScopesContains(oidcClient, requestAudienceScopeName) && !allowedGrantTypesContains(oidcClient, tokenExchangeGrantTypeName):
+		conditions = append(conditions, &v1alpha1.Condition{
+			Type:   allowedGrantTypesValid,
+			Status: v1alpha1.ConditionFalse,
+			Reason: reasonMissingRequiredValue,
+			Message: fmt.Sprintf("%q must be included in %q when %q is included in %q",
+				tokenExchangeGrantTypeName, allowedGrantTypesFieldName, requestAudienceScopeName, allowedScopesFieldName),
+		})
+	default:
+		conditions = append(conditions, &v1alpha1.Condition{
+			Type:    allowedGrantTypesValid,
+			Status:  v1alpha1.ConditionTrue,
+			Reason:  reasonSuccess,
+			Message: fmt.Sprintf("%q is valid", allowedGrantTypesFieldName),
+		})
+	}
+	return conditions
+}
+
+// validateSecret checks if the client secret storage Secret is valid and contains at least one client secret.
+func validateSecret(secret *v1.Secret, conditions []*v1alpha1.Condition) []*v1alpha1.Condition {
+	if secret == nil {
+		// Invalid: no storage Secret found.
+		conditions = append(conditions, &v1alpha1.Condition{
+			Type:    clientSecretExists,
+			Status:  v1alpha1.ConditionFalse,
+			Reason:  reasonNoClientSecretFound,
+			Message: "no client secret found (no Secret storage found)",
+		})
+		return conditions
+	}
+
+	storedClientSecret, err := oidcclientsecretstorage.ReadFromSecret(secret)
+	if err != nil {
+		// Invalid: storage Secret exists but its data could not be parsed.
+		conditions = append(conditions, &v1alpha1.Condition{
+			Type:    clientSecretExists,
+			Status:  v1alpha1.ConditionFalse,
+			Reason:  reasonNoClientSecretFound,
+			Message: fmt.Sprintf("error reading client secret storage: %s", err.Error()),
+		})
+		return conditions
+	}
+
+	// Successfully read the stored client secrets, so check if there are any stored in the list.
+	storedClientSecretsCount := len(storedClientSecret.SecretHashes)
+	if storedClientSecretsCount == 0 {
```

Comment:
This is supposed to be part of the status (as a distinct field + column in the table output).

Reply:
Updated in #1221.